stage0/kvm.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stage0
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+const (
+	kvmSettingsDir        = "/kvm"
+	kvmPrivateKeyFilename = "/ssh_kvm_key"
+)
+
+func fileAccessible(path string) bool {
+	info, err := os.Stat(path)
+	if err == nil {
+		return info.Mode().IsRegular()
+	}
+	return false
+}
+
+func sshPrivateKeyPath(dataDir string) string {
+	return filepath.Join(dataDir, kvmSettingsDir, kvmPrivateKeyFilename)
+}
+
+func sshPublicKeyPath(dataDir string) string {
+	return filepath.Join(dataDir, kvmSettingsDir, kvmPrivateKeyFilename+".pub")
+}
+
+func ensureDirExists(path string) error {
+	err := os.MkdirAll(path, 0700)
+	if err != nil && os.IsExist(err) {
+		return nil
+	}
+	return err
+}
+
+func createKvmSettingsDir(dataDir string) error {
+	return ensureDirExists(filepath.Join(dataDir, kvmSettingsDir))
+}
+
+func generateKeyPair(private string) error {
+	out, err := exec.Command("ssh-keygen", "-q", "-t", "dsa", "-b", "1024", "-f", private).Output()
+	if err != nil {
+		fmt.Printf("Error in keygen time. Output: %v\n", string(out[:len(out)]))
+	}
+	return err
+}
+
+func ensureKeysExistenceOnHost(dataDir string) error {
+	// check existance of files
+	private, public := sshPrivateKeyPath(dataDir), sshPublicKeyPath(dataDir)
+	if !fileAccessible(private) || !fileAccessible(public) {
+		err := createKvmSettingsDir(dataDir)
+		if err != nil {
+			return err
+		}
+
+		err = generateKeyPair(private)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ensureAuthorizedKeysExist(keyDirPath, dataDir string) (err error) {
```

Comment:
This wasn't addressed.

To avoid using named error value, you can split it into two functions, where one calls another one:
```go
func outer(...) error {
	fout, err := os.OpenFile(...); if err != nil {return err}
	err := inner(fout, ...)
	cerr := fout.Close()
	if err == nil {return cerr}
	return err
}

func inner(fout io.WriterOrSomethingMoreSpecific, ...) error {
	fin, err := os.Open(...); if err != nil {return err}
	defer fin.Close()
	if _, err = io.Copy(fout, fin); err != nil {return err}
	return fout.Sync()
}
```